platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/AbstractCollectionProperty.java

```diff
@@ -196,11 +196,8 @@ public int size() {
      */
     private void addExplicitCollector(Collector<T> collector, boolean ignoreAbsent) {
         assertCanMutate();
-        CollectionSupplier<T, C> explicitValue = getExplicitValue(defaultValue);
-        if (ignoreAbsent) {
-            explicitValue = explicitValue.ignoringAbsent();
-        }
-        setSupplier(explicitValue.plus(collector));
+        CollectionSupplier<T, C> explicitValue = getExplicitValue(defaultValue).absentIgnoring(ignoreAbsent);
+        setSupplier(explicitValue.plus(collector.absentIgnoring(ignoreAbsent)));
```

Comment:
❓ Should we also call `absentIgnoring(ignoreAbsent)` on the result of `plus`? It looks like otherwise we'll end up with a `CollectingSupplier` that has `false` as the `absentIgnoring`. Something like
`myList.append(providers.environmentVariable("FOO"))` is going to be poisoned if the `FOO` isn't set after CC roundtrip.

Actually, even this won't help, because the roundtrip will erase all `Supplier` decorators, and we'll end up with a `CollectingProvider` that cannot tell which elements are safe.

Reply:
I *believe* I got this right now. I am doing it on the resulting PlusCollector though.  To keep the implementation of PlusCollector simple, instead of making it handle both cases (where its result could be absent or not), I created a separate class to model the case it should never return absent.